src/contracts/atlas/Escrow.sol

```diff
@@ -43,18 +47,95 @@ contract Escrow is DAppVerification, SafetyLocks, FastLaneErrorsEvents {
+        }
+
+        emit Transfer(msg.sender, to, amount);
+
+        return true;
+    }
+
+    function transferFrom(address from, address to, uint256 amount) public override returns (bool) {
+        uint256 allowed = allowance[from][msg.sender]; // Saves gas for limited approvals.
+
+        if (allowed != type(uint256).max) allowance[from][msg.sender] = allowed - amount;
+
+        unlockedBalanceOf[from] -= amount;
+
+        // Cannot overflow because the sum of all user
+        // balances can't exceed the max uint256 value.
+        unchecked {
+            unlockedBalanceOf[to] += amount;
+        }
+
+        emit Transfer(from, to, amount);
+
+        return true;
     }
 
     ///////////////////////////////////////////////////
     /// EXTERNAL FUNCTIONS FOR SOLVER INTERACTION ///
     ///////////////////////////////////////////////////
-    function deposit(address solverSigner) onlyWhenUnlocked external payable returns (uint256 newBalance) {
-        // NOTE: The escrow accounting system cannot currently handle deposits made mid-transaction.
 
-        _escrowData[solverSigner].total += uint128(msg.value);
-        newBalance = uint256(_escrowData[solverSigner].total);
+    // Deposit ETH and get AETH in return. The minted AETH amount is added to the unlocked balance,
+    // meaning it can be freely transferred or withdrawn, but won't be available for escrow.
+    // After depositing, call lock() to escrow the AETH and begin bidding.
+    function deposit() external payable returns (uint256 newUnlockedBalance, uint256 newBalance) {
+        _mint(msg.sender, msg.value);
+        unlockedBalanceOf[msg.sender] += msg.value;
+        newUnlockedBalance = unlockedBalanceOf[msg.sender];
+        newBalance = balanceOf[msg.sender];
+    }
+
+    // Redeem AETH for ETH. Only unlocked (non-escrowed) AETH can be redeemed.
+    function withdraw(uint256 amount) external returns (uint256 newUnlockedBalance, uint256 newBalance) {
+        require(unlockedBalanceOf[msg.sender] >= amount, "ERR-E078 InsufficientUnlockedBalance");
+        _burn(msg.sender, amount);
+        SafeTransferLib.safeTransferETH(msg.sender, amount);
+        unlockedBalanceOf[msg.sender] -= amount;
+        newUnlockedBalance = unlockedBalanceOf[msg.sender];
+        newBalance = balanceOf[msg.sender];
+    }
+
+    // Lock AETH in escrow.
+    function lock(uint256 amount) external onlyWhenUnlocked returns (uint256 newLockedBalance) {
+        require(unlockedBalanceOf[msg.sender] >= amount, "ERR-E079 InsufficientUnlockedBalance");
+        unlockedBalanceOf[msg.sender] -= amount;
+        _escrowData[msg.sender].total += amount;
+        newLockedBalance = uint256(_escrowData[msg.sender].total);
+    }
+
+    // Unlock AETH from escrow. Only eligible after the escrowDuration from the last solver's
+    // interaction with the contract has elapsed.
+    function unlock(uint256 amount) external onlyWhenUnlocked returns (uint256 newLockedBalance) {
```

Comment:
The lock/unlock term is a bit confusing when also referring to the activeEnv lock. E.g. in this function it looks like I can only call `unlock` (to unlock my AETH) when it is already unlocked (because of the `onlyWhenUnlocked` modifier that refers to the activeEnv lock)

Maybe we can have different names for those concepts

Reply:
Agreed. Replaced lock/unlock by escrowBalance and unescrowBalance.